pulsar-broker/src/main/java/org/apache/pulsar/compaction/PulsarTopicCompactionService.java

```diff
@@ -106,6 +111,78 @@ public CompletableFuture<Position> getLastCompactedPosition() {
         return CompletableFuture.completedFuture(compactedTopic.getCompactionHorizon().orElse(null));
     }
 
+    @Override
+    public CompletableFuture<Entry> findEntryByPublishTime(long publishTime) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                return Commands.parseMessageMetadata(entry.getDataBuffer()).getPublishTime() >= publishTime;
+            } finally {
+                entry.release();
+            }
+        };
+        return findFirstMatchEntry(predicate);
+    }
+
+    @Override
+    public CompletableFuture<Entry> findEntryByEntryIndex(long entryIndex) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                BrokerEntryMetadata brokerEntryMetadata =
+                        Commands.parseBrokerEntryMetadataIfExist(entry.getDataBuffer());
+                if (brokerEntryMetadata == null || !brokerEntryMetadata.hasIndex()) {
+                    return false;
+                }
+                return brokerEntryMetadata.getIndex() >= entryIndex;
+            } finally {
+                entry.release();
```

Comment:
Why we can't move the release to `findFirstMatchIndexLoop` method?

Reply:
The `release` can be moved to `findFirstMatchIndexLoop` method, I improve it.

Reply:
Well. I suggest it because I think we don't need to release it everywhere. Do you agree with this idea?

Reply:
Yes, the code will be unified